### DISCUSSION TITLE
Getting "Method not found: Amazon.Runtime.Internal.Util.ChecksumUtils.SetChecksumData" error while uploading file on S3

### DISCUSSION THREAD

**Discussion by jaiminvs-devit** [aws/aws-sdk-net#3124] [Q&A] [answered]
### Describe the bug

I am trying to upload file on S3 bucket using C# and AWS Single Sign-On (SSO) end-user access and my code works for .NET Standard 7.0 but when I write very same code in .NET Framework 4.7.2, I get below message:
Method not found: 'Void Amazon.Runtime.Internal.Util.ChecksumUtils.SetChecksumData(Amazon.Runtime.Internal.IRequest, System.String, Boolean)'.

also I have tried with both PutObject and PutObjectAsync functions but the exception remains the same.
I have attacked snippets of my code, exception details and reference added into my project.
![Exception Details](https://github.com/aws/aws-sdk-net/assets/55880096/00bc0324-a135-4d4c-a8e0-c317aa6ce66b)
![References added](https://github.com/aws/aws-sdk-net/assets/55880096/868d9cba-61b1-4ddb-9a6d-237e80da76f5)

### Expected Behavior

As of my understanding, SetChecksumData method is called during internal operations. In such cases no such exceptions should have been thrown and the file should have been uploaded.

### Current Behavior

Getting below message instead of file being uploaded:
Method not found: 'Void Amazon.Runtime.Internal.Util.ChecksumUtils.SetChecksumData(Amazon.Runtime.Internal.IRequest, System.String, Boolean)'.

### Reproduction Steps
```C#
string bucketName = "<Bucket>";
string objectName = Path.GetFileName(filePath);
AmazonS3Config config = new AmazonS3Config();
Amazon.Profile profile = new Amazon.Profile("<Profile>");
config.Profile = profile;
config.RegionEndpoint = AWS_SSM_Helper.S3_Region;
using (var client = new AmazonS3Client(config))
{
    var request = new PutObjectRequest
    {
        BucketName = bucketName,
        Key = keyName,
        InputStream = inputStream
    };

    var response = await client.PutObjectAsync(request);
    //if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
    //{
    //    Console.WriteLine($"Successfully uploaded {objectName} to {bucketName}.");
    return true;
    //}
    //else
    //{
    //    Console.WriteLine($"Could not upload {objectName} to {bucketName}.");
    //    return false;
    //}
}
```

### AWS .NET SDK and/or Package version used

AWSSDK.Core - 3.7.300.12
AWSSDK.S3 - 3.7.304.1
AWSSDK.SimpleSystemsManagement - 3.7.301.8
AWSSDK.SSO - 3.7.300.12
AWSSDK.SSOOIDC - 3.7.301.7

### Targeted .NET Platform

.NET Framework 4.7.2

### Operating System and version

Windows 10 Enterprise

**Comment by ashishdhingra** [comment]
@dev2-devitpl Good morning. Thanks for reporting the issue.

In the commit history for [ChecksumUtils](https://github.com/aws/aws-sdk-net/commits/main/sdk/src/Core/Amazon.Runtime/Internal/Util/ChecksumUtils.cs), the [commit](https://github.com/aws/aws-sdk-net/commit/8bf2a7ce6ba30ba74e9897d8b21c3963ebe31bce) added public `SetChecksumData()` method. It is present in the latest commit [here](https://github.com/aws/aws-sdk-net/blob/f7902bffb88dcafbafc1913d7a786ebab4d84496/sdk/src/Core/Amazon.Runtime/Internal/Util/ChecksumUtils.cs#L266).

The [first commit](https://github.com/aws/aws-sdk-net/commit/4a57ae704b2cee415ea4ad59a468f60f60ca5c0c) which didn't include this method was released in `AWSSDK.Core` version `3.7.8.0` per [ChangeLog entry](https://github.com/aws/aws-sdk-net/blob/main/changelogs/SDK.CHANGELOG.2022.md#372140-2022-02-22-1438-utc).

Since you migrated from NetStandard to `.NET Framework 4.7.2`, you might have some floating old assemblies around. Could you confirm the following:
- Update all `AWSSDK.*` packages to the latest version, clean the solution and see if the error goes away?
- Do you have by any chance AWSSDK assemblies installed in GAC? GAC always take precedence over local assemblies and hence it is not a good idea to install AWSSDK assemblies in GAC since due to legacy reasons, the assembly version is fixed at `3.3.0.0`. So it would always load assembly from GAC, if it is present.

Thanks,
Ashish

**Comment by jaiminvs-devit** [comment]
Hello Ashish, Thank you for prompt response. I did perform the steps you suggested but it did not work for me. also I am using Nuget Packages so I think there might be some issue with them.

**Comment by ashishdhingra** [comment]
> Hello Ashish, Thank you for prompt response. I did perform the steps you suggested but it did not work for me. also I am using Nuget Packages so I think there might be some issue with them.

@dev2-devitpl Good afternoon. Somehow I'm unable to reproduce the issue at my end. Below is my `.csproj` file for your reference:
```XML
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{22463005-CBE7-441E-A79A-2A677D2C8AF7}</ProjectGuid>
    <OutputType>Exe</OutputType>
    <RootNamespace>S3TestNetFramework472</RootNamespace>
    <AssemblyName>S3TestNetFramework472</AssemblyName>
    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
    <FileAlignment>512</FileAlignment>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <Deterministic>true</Deterministic>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugSymbols>true</DebugSymbols>
    <DebugType>full</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugType>pdbonly</DebugType>
    <Optimize>true</Optimize>
    <OutputPath>bin\Release\</OutputPath>
    <DefineConstants>TRACE</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
  </PropertyGroup>
  <ItemGroup>
    <Reference Include="AWS.EncryptionSDK, Version=3.1.0.0, Culture=neutral, processorArchitecture=MSIL">
      <HintPath>..\packages\AWS.EncryptionSDK.3.1.0\lib\net452\AWS.EncryptionSDK.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.Core, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.Core.3.7.300.14\lib\net45\AWSSDK.Core.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.KeyManagementService, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.KeyManagementService.3.7.300.14\lib\net45\AWSSDK.KeyManagementService.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.S3, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.S3.3.7.304.3\lib\net45\AWSSDK.S3.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.SimpleSystemsManagement, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.SimpleSystemsManagement.3.7.301.10\lib\net45\AWSSDK.SimpleSystemsManagement.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.SSO, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.SSO.3.7.300.14\lib\net45\AWSSDK.SSO.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.SSOOIDC, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.SSOOIDC.3.7.301.9\lib\net45\AWSSDK.SSOOIDC.dll</HintPath>
    </Reference>
    <Reference Include="AWSSDK.Transfer, Version=3.3.0.0, Culture=neutral, PublicKeyToken=885c28607f98e604, processorArchitecture=MSIL">
      <HintPath>..\packages\AWSSDK.Transfer.3.7.301.10\lib\net45\AWSSDK.Transfer.dll</HintPath>
    </Reference>
    <Reference Include="BouncyCastle.Crypto, Version=1.8.5.0, Culture=neutral, PublicKeyToken=0e99375e54769942, processorArchitecture=MSIL">
      <HintPath>..\packages\Portable.BouncyCastle.1.8.5.2\lib\net40\BouncyCastle.Crypto.dll</HintPath>
    </Reference>
    <Reference Include="System" />
    <Reference Include="System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Buffers.4.4.0\lib\netstandard2.0\System.Buffers.dll</HintPath>
    </Reference>
    <Reference Include="System.Collections.Immutable, Version=1.2.5.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Collections.Immutable.1.7.0\lib\netstandard2.0\System.Collections.Immutable.dll</HintPath>
    </Reference>
    <Reference Include="System.Core" />
    <Reference Include="System.Memory, Version=4.0.1.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Memory.4.5.3\lib\netstandard2.0\System.Memory.dll</HintPath>
    </Reference>
    <Reference Include="System.Numerics" />
    <Reference Include="System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Numerics.Vectors.4.4.0\lib\net46\System.Numerics.Vectors.dll</HintPath>
    </Reference>
    <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.4.5.2\lib\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
    </Reference>
    <Reference Include="System.ValueTuple, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51, processorArchitecture=MSIL">
      <HintPath>..\packages\System.ValueTuple.4.5.0\lib\net47\System.ValueTuple.dll</HintPath>
    </Reference>
    <Reference Include="System.Xml.Linq" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="Microsoft.CSharp" />
    <Reference Include="System.Data" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Xml" />
  </ItemGroup>
  <ItemGroup>
    <Compile Include="Program.cs" />
    <Compile Include="Properties\AssemblyInfo.cs" />
  </ItemGroup>
  <ItemGroup>
    <None Include="App.config" />
    <None Include="packages.config" />
  </ItemGroup>
  <ItemGroup>
    <Analyzer Include="..\packages\AWSSDK.KeyManagementService.3.7.300.14\analyzers\dotnet\cs\AWSSDK.KeyManagementService.CodeAnalysis.dll" />
    <Analyzer Include="..\packages\AWSSDK.S3.3.7.304.3\analyzers\dotnet\cs\AWSSDK.S3.CodeAnalysis.dll" />
    <Analyzer Include="..\packages\AWSSDK.S3.3.7.304.3\analyzers\dotnet\cs\SharedAnalysisCode.dll" />
    <Analyzer Include="..\packages\AWSSDK.SimpleSystemsManagement.3.7.301.10\analyzers\dotnet\cs\AWSSDK.SimpleSystemsManagement.CodeAnalysis.dll" />
    <Analyzer Include="..\packages\AWSSDK.SSO.3.7.300.14\analyzers\dotnet\cs\AWSSDK.SSO.CodeAnalysis.dll" />
    <Analyzer Include="..\packages\AWSSDK.SSOOIDC.3.7.301.9\analyzers\dotnet\cs\AWSSDK.SSOOIDC.CodeAnalysis.dll" />
    <Analyzer Include="..\packages\AWSSDK.Transfer.3.7.301.10\analyzers\dotnet\cs\AWSSDK.Transfer.CodeAnalysis.dll" />
  </ItemGroup>
  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
</Project>
```

I suspect you might have `AWSSDK.Core` assembly in GAC. Kindly execute `gacutil /l` to list the assemblies in GAC and see the output if you have older version of `AWSSDK.*` assembly in GAC.

Thanks,
Ashish

**Comment by jaiminvs-devit** [comment]
above issue is resolved but I am getting another error in VSTA Project. Error looks something like **'..\packages\AWSSDK.SimpleSystemsManagement.3.7.202.4\analyzers\dotnet\cs\AWSSDK.SimpleSystemsManagement.CodeAnalysis.dll' could not be found**. any suggestions?

**Comment by ashishdhingra** [comment, marked as answer]
> above issue is resolved but I am getting another error in VSTA Project. Error looks something like **'..\packages\AWSSDK.SimpleSystemsManagement.3.7.202.4\analyzers\dotnet\cs\AWSSDK.SimpleSystemsManagement.CodeAnalysis.dll' could not be found**. any suggestions?

@dev2-devitpl The package [AWSSDK.SimpleSystemsManagement](https://nuget.info/packages/AWSSDK.SimpleSystemsManagement/3.7.301.10) does include the `AWSSDK.SimpleSystemsManagement.CodeAnalysis.dll` assembly. Below is the screenshot for my project:
![Screenshot 2023-12-06 at 11 46 01 AM](https://github.com/aws/aws-sdk-net/assets/67916761/36fd4aac-1163-47e0-bdc1-e5b35c69db56)

- What does your project structure looks like?
- Could you check the path `..\packages\AWSSDK.SimpleSystemsManagement.3.7.301.10\analyzers\dotnet\cs\AWSSDK.SimpleSystemsManagement.CodeAnalysis.dll` relative to your VSTA project in case you copy-pasted from the `.csproj` I shared earlier? This is more like issue with your project structure, not related to SDK.

Thanks,
Ashish

**Comment by jaiminvs-devit** [comment]
Hello Ashish, I was able to upload file to S3 bucket from console app so that was a success. thank you :)
Now to the main part, I am trying to upload file to S3 bucket from SSIS Package using SSO. I had to add all DLLs to GAC as DLLs could not be added as reference.
![image](https://github.com/aws/aws-sdk-net/assets/55880096/e9cd54f3-b687-40db-91c8-dde8d635409d)


I tried different tactics to resolve this issue but can't proceed further. Suggestions?

**Reply by ashishdhingra** [reply]
@dev2-devitpl Please refer section `Workaround if you cannot install the Assembly in the GAC` in https://blog.matrixpost.net/load-an-custom-assembly-in-sql-server-integration-services-ssis-script-task-that-is-not-or-cannot-be-stored-in-the-gac/. It might work. Kindly test it in non-prod environment first; we are just posting this link for guidance.

**Comment by ashishdhingra** [comment]
Moving this issue to discussion since this is not the issue with SDK, but user is requesting general guidance.
